### PR TITLE
Add Mojo::Promise handling to the eval command

### PR DESCRIPTION
**Pull request by jberger**
I find the eval command to be invaluable for one-off usage and testing, especially while developing. As I've been writing more promise apis and now as I'm writing fewer blocking apis due to the existence of Mojo::AsyncAwait, I'm finding the usage of promises in the eval command to be cumbersome to the point of not using it. With this change, one that I doubt would affect any usage in the wild (who wants to dump the stringified promise?), we can make a very comfortable interface for one-off testing of promise apis and make for easy examples and demos for newcomers or answer-seekers on irc.

### CODE REVIEW

File: lib/Mojolicious/Command/eval.pm [hunk 1]
```diff
@@ -49,6 +56,7 @@ Mojolicious::Command::eval - Eval command
 =head1 DESCRIPTION
 
 L<Mojolicious::Command::eval> runs code against applications.
+If the result is a L<Mojo::Promise>, it will wait until the promise is completed and the result is returned.
```
Comment (jberger) [inline lib/Mojolicious/Command/eval.pm]: Before the PR was opened @kraih commented on this line on a commit https://github.com/mojolicious/mojo/commit/4eaf550044e988635d41f4ce0349a2c18acf6b9d#r31802449

Comment (jberger) [reply lib/Mojolicious/Command/eval.pm]: Do you mean just my added sentence or the existing one as well?

Comment (jberger) [reply lib/Mojolicious/Command/eval.pm]: Oh! Do you just mean the literal line formatting? Yeah, I'll fix that, sorry.

File: lib/Mojolicious/Command/eval.pm [hunk 2]
```diff
@@ -16,7 +17,13 @@ sub run {
   my $app = $self->app;
   no warnings;
   my $result = eval "package main; sub app; local *app = sub { \$app }; $code";
-  return $@ ? die $@ : $result unless defined $result && ($v1 || $v2);
+  die $@ if $@;
+  if (blessed $result && $result->isa('Mojo::Promise')) {
```
Comment (jhthorsen) [inline lib/Mojolicious/Command/eval.pm]: Would it be to generic to use the same logic as inside Mojo::Promise instead?

https://github.com/mojolicious/mojo/blob/fb702a7a7618c82fff57b1a0d7d49ac71eb8ee7d/lib/Mojo/Promise.pm#L97

(Guess the documentation has to be updated as well then)

**Comment by kraih** [comment]
I never had a need for this, so not sure how to vote.

File: lib/Mojolicious/Command/eval.pm [hunk 3]
```diff
@@ -48,7 +55,9 @@ Mojolicious::Command::eval - Eval command
 
 =head1 DESCRIPTION
 
-L<Mojolicious::Command::eval> runs code against applications.
+L<Mojolicious::Command::eval> runs code against applications. If the result is
+a L<Mojo::Promise>, it will wait until the promise is completed and the result
```
Comment (kraih) [inline lib/Mojolicious/Command/eval.pm]: The documentation for `Mojo::Promise::wait` says "fulfilled or rejected" instead of "completed".

Comment (jberger) [reply lib/Mojolicious/Command/eval.pm]: I can make this change the next time I'm at a computer

**Comment by jberger** [comment]
> I never had a need for this, so not sure how to vote.

I had only rarely until a new project at $work. Since async functions necessarily return promises and since they nearly obviate the need for blocking apis I now have an app for which the model is entirely promises. I think once aa becomes more common this will be the case more and more often, it feels very natural

File: lib/Mojolicious/Command/eval.pm (hunk 2 above)
Comment (jberger) [reply lib/Mojolicious/Command/eval.pm]: As of the second commit in this PR, arbitrary then-ables are now handled.

**Comment by jberger** [comment]
@marcusramberg I know you're on vacation, but if you get a second, I'd appreciate a review/vote

**Comment by jberger** [comment]
Per the new PR policy, I'm calling a vote on this PR. @mojolicious/core

**Comment by kraih** [comment]
Neutral vote from me.

**Comment by marcusramberg** [comment]
I'm also in favor of this change 👍

**Comment by s1037989** [comment]
What do you think about adding a Usage example?

```
Usage: APPLICATION eval [OPTIONS] CODE

  ./myapp.pl eval 'say app->ua->get("/")->result->body'
  ./myapp.pl eval 'say for sort keys %{app->renderer->helpers}'
  ./myapp.pl eval -v 'app->home'
  ./myapp.pl eval -V 'app->renderer->paths'
  ./myapp.pl eval -V '
```

I don't know if this is a good example, but something like:
```
  ./myapp.pl eval -v 'app->ua->get_p("https://mojolicious.org")->then(sub{shift->result->dom->at("title")->text})'
```

**Pull request merged**